File: ui-tests/Utils/RetryHelpers.cs
using System;
using System.Threading.Tasks;

namespace UiTests.Utils;

/// <summary>
/// Helper utilities for retrying flaky asynchronous operations.
/// </summary>
public static class RetryHelpers
{
    private const int DetailedAttemptLimit = 3;
    private const int SummaryBatchSize = 5;

    /// <summary>
    /// Repeatedly executes an asynchronous condition until it returns <c>true</c>
    /// or the retry count is exhausted.
    /// </summary>
    public static async Task RetryAsync(Func<Task<bool>> condition, int maxAttempts = 10, int delayMs = 1000)
    {
        var loggingEnabled = DebugLogger.IsEnabled;
        RetrySuppressionState? suppression = null;

        if (loggingEnabled)
        {
            DebugLogger.Log($"RetryAsync<bool>: start (maxAttempts={maxAttempts}, delayMs={delayMs})");
        }

        for (int attempt = 1; attempt <= maxAttempts; attempt++)
        {
            if (loggingEnabled && attempt <= DetailedAttemptLimit)
            {
                DebugLogger.Log($"RetryAsync<bool>: attempt {attempt} evaluating condition");
            }

            if (await condition())
            {
                if (loggingEnabled)
                {
                    suppression?.Flush("bool");
                    DebugLogger.Log($"RetryAsync<bool>: attempt {attempt} succeeded");
                }
                return;
            }

            var willRetry = attempt < maxAttempts;
            if (loggingEnabled)
            {
                if (attempt <= DetailedAttemptLimit)
                {
                    var suffix = willRetry ? $"; sleeping {delayMs}ms" : string.Empty;
                    DebugLogger.Log($"RetryAsync<bool>: attempt {attempt} failed{suffix}");
                }
                else
                {
                    suppression ??= new RetrySuppressionState(SummaryBatchSize);
                    suppression.RecordFailure("bool", attempt);
                }
            }

            if (willRetry)
            {
                await Task.Delay(delayMs);
            }
        }

        if (loggingEnabled)
        {
            suppression?.Flush("bool");
            DebugLogger.Log($"RetryAsync<bool>: exhausted {maxAttempts} attempts without success");
        }

        throw new TimeoutException($"Condition was not satisfied after {maxAttempts} attempts.");
    }

    /// <summary>
    /// Repeatedly invokes an asynchronous action until it completes without
    /// throwing an exception or the retry count is exceeded.
    /// </summary>
    public static async Task RetryAsync(Func<Task> action, int maxAttempts = 50, int delayMs = 100)
    {
        Exception? lastError = null;
        var loggingEnabled = DebugLogger.IsEnabled;
        RetrySuppressionState? suppression = null;

        if (loggingEnabled)
        {
            DebugLogger.Log($"RetryAsync<void>: start (maxAttempts={maxAttempts}, delayMs={delayMs})");
        }

        for (int attempt = 1; attempt <= maxAttempts; attempt++)
        {
            try
            {
                if (loggingEnabled && attempt <= DetailedAttemptLimit)
                {
                    DebugLogger.Log($"RetryAsync<void>: attempt {attempt} executing action");
                }

                await action();

                if (loggingEnabled)
                {
                    suppression?.Flush("void");
                    DebugLogger.Log($"RetryAsync<void>: attempt {attempt} completed");
                }

                return;
            }
            catch (Exception ex)
            {
                lastError = ex;
                if (loggingEnabled)
                {
                    if (attempt <= DetailedAttemptLimit)
                    {
                        DebugLogger.Log($"RetryAsync<void>: attempt {attempt} threw {ex.GetType().Name}: {ex.Message}");
                    }
                    else
                    {
                        suppression ??= new RetrySuppressionState(SummaryBatchSize);
                        suppression.RecordFailure("void", attempt);
                    }
                }
            }

            var willRetry = attempt < maxAttempts;
            if (!willRetry)
            {
                break;
            }

            if (loggingEnabled && attempt <= DetailedAttemptLimit)
            {
                DebugLogger.Log($"RetryAsync<void>: sleeping {delayMs}ms before next attempt");
            }

            await Task.Delay(delayMs);
        }

        if (loggingEnabled)
        {
            suppression?.Flush("void");
            DebugLogger.Log($"RetryAsync<void>: exhausted {maxAttempts} attempts; throwing TimeoutException");
        }

        throw new TimeoutException($"Action failed after {maxAttempts} attempts", lastError);
    }

    private sealed class RetrySuppressionState
    {
        private readonly int _summaryBatchSize;
        private int _rangeStart = -1;
        private int _count;

        public RetrySuppressionState(int summaryBatchSize)
        {
            _summaryBatchSize = summaryBatchSize;
        }

        public void RecordFailure(string label, int attempt)
        {
            if (_rangeStart < 0)
            {
                _rangeStart = attempt;
            }

            _count++;

            if (_count >= _summaryBatchSize)
            {
                Flush(label);
            }
        }

        public void Flush(string label)
        {
            if (_count == 0 || _rangeStart < 0)
            {
                _rangeStart = -1;
                _count = 0;
                return;
            }

            var end = _rangeStart + _count - 1;
            var descriptor = _rangeStart == end ? $"attempt {end}" : $"attempts {_rangeStart}-{end}";
            DebugLogger.Log($"RetryAsync<{label}>: {descriptor} failed; continuing retries");
            _rangeStart = -1;
            _count = 0;
        }
    }
}

File: tsc-ui-tests/lib/retry-helpers.ts
import { debugLogger } from "./debug-logger";

const DETAILED_ATTEMPT_LIMIT = 3;
const SUMMARY_BATCH_SIZE = 5;

interface RetryOptions {
  /** Maximum number of attempts before giving up. */
  maxAttempts?: number;
  /** Delay in milliseconds between attempts. */
  delayMs?: number;
}

/**
 * Batched logging state for retry attempts beyond the detailed limit.
 * Emits a summary line every `batchSize` failures instead of one per attempt.
 */
class RetrySuppressionState {
  private rangeStart = -1;
  private count = 0;

  constructor(private readonly batchSize: number) {}

  recordFailure(label: string, attempt: number): void {
    if (this.rangeStart < 0) {
      this.rangeStart = attempt;
    }
    this.count++;
    if (this.count >= this.batchSize) {
      this.flush(label);
    }
  }

  flush(label: string): void {
    if (this.count === 0 || this.rangeStart < 0) {
      this.rangeStart = -1;
      this.count = 0;
      return;
    }

    const end = this.rangeStart + this.count - 1;
    const descriptor =
      this.rangeStart === end
        ? `attempt ${end}`
        : `attempts ${this.rangeStart}-${end}`;
    debugLogger.log(`RetryAsync<${label}>: ${descriptor} failed; continuing retries`);
    this.rangeStart = -1;
    this.count = 0;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Repeatedly evaluates an async condition until it returns `true`
 * or the attempt limit is exhausted.
 *
 * Prefer Playwright's `await expect(async () => { ... }).toPass()` for
 * assertion-style waits. Use `retry()` for non-assertion waits where
 * you need to poll until a side-effect completes.
 *
 * Port of ui-tests/Utils/RetryHelpers.cs `RetryAsync(Func<Task<bool>>)`.
 */
export async function retry(
  condition: () => Promise<boolean>,
  opts?: RetryOptions,
): Promise<void> {
  const maxAttempts = opts?.maxAttempts ?? 10;
  const delayMs = opts?.delayMs ?? 1000;
  const logging = debugLogger.isEnabled;
  let suppression: RetrySuppressionState | null = null;

  if (logging) {
    debugLogger.log(
      `RetryAsync<bool>: start (maxAttempts=${maxAttempts}, delayMs=${delayMs})`,
    );
  }

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    if (logging && attempt <= DETAILED_ATTEMPT_LIMIT) {
      debugLogger.log(`RetryAsync<bool>: attempt ${attempt} evaluating condition`);
    }

    if (await condition()) {
      if (logging) {
        suppression?.flush("bool");
        debugLogger.log(`RetryAsync<bool>: attempt ${attempt} succeeded`);
      }
      return;
    }

    const willRetry = attempt < maxAttempts;
    if (logging) {
      if (attempt <= DETAILED_ATTEMPT_LIMIT) {
        const suffix = willRetry ? `; sleeping ${delayMs}ms` : "";
        debugLogger.log(`RetryAsync<bool>: attempt ${attempt} failed${suffix}`);
      } else {
        suppression ??= new RetrySuppressionState(SUMMARY_BATCH_SIZE);
        suppression.recordFailure("bool", attempt);
      }
    }

    if (willRetry) {
      await sleep(delayMs);
    }
  }

  if (logging) {
    suppression?.flush("bool");
    debugLogger.log(
      `RetryAsync<bool>: exhausted ${maxAttempts} attempts without success`,
    );
  }

  throw new Error(`Condition was not satisfied after ${maxAttempts} attempts.`);
}

/**
 * Repeatedly invokes an async action until it completes without throwing,
 * or the attempt limit is exhausted.
 *
 * Port of ui-tests/Utils/RetryHelpers.cs `RetryAsync(Func<Task>)`.
 */
export async function retryAction(
  action: () => Promise<void>,
  opts?: RetryOptions,
): Promise<void> {
  const maxAttempts = opts?.maxAttempts ?? 50;
  const delayMs = opts?.delayMs ?? 100;
  const logging = debugLogger.isEnabled;
  let suppression: RetrySuppressionState | null = null;
  let lastError: Error | null = null;

  if (logging) {
    debugLogger.log(
      `RetryAsync<void>: start (maxAttempts=${maxAttempts}, delayMs=${delayMs})`,
    );
  }

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      if (logging && attempt <= DETAILED_ATTEMPT_LIMIT) {
        debugLogger.log(`RetryAsync<void>: attempt ${attempt} executing action`);
      }

      await action();

      if (logging) {
        suppression?.flush("void");
        debugLogger.log(`RetryAsync<void>: attempt ${attempt} completed`);
      }

      return;
    } catch (ex) {
      lastError = ex instanceof Error ? ex : new Error(String(ex));
      if (logging) {
        if (attempt <= DETAILED_ATTEMPT_LIMIT) {
          debugLogger.log(
            `RetryAsync<void>: attempt ${attempt} threw ${lastError.name}: ${lastError.message}`,
          );
        } else {
          suppression ??= new RetrySuppressionState(SUMMARY_BATCH_SIZE);
          suppression.recordFailure("void", attempt);
        }
      }
    }

    const willRetry = attempt < maxAttempts;
    if (!willRetry) {
      break;
    }

    if (logging && attempt <= DETAILED_ATTEMPT_LIMIT) {
      debugLogger.log(
        `RetryAsync<void>: sleeping ${delayMs}ms before next attempt`,
      );
    }

    await sleep(delayMs);
  }

  if (logging) {
    suppression?.flush("void");
    debugLogger.log(
      `RetryAsync<void>: exhausted ${maxAttempts} attempts; throwing`,
    );
  }

  throw new Error(
    `Action failed after ${maxAttempts} attempts`,
    lastError ? { cause: lastError } : undefined,
  );
}
